diff --git a/quiz.admin.inc b/quiz.admin.inc
index 50d99e7..d0ddcc3 100644 (file)
--- a/quiz.admin.inc
+++ b/quiz.admin.inc
@@ -1065,7 +1065,7 @@ function _quiz_add_questions_to_form(&$form, &$questions, &$quiz, &$question_typ
         '#type' => 'textfield',
         '#size' => 2,
         '#maxlength' => 2,
-        '#disabled' => $question->auto_update_max_score,
+        '#disabled' => isset($question->auto_update_max_score) ? $question->auto_update_max_score : FALSE,
         '#default_value' => isset($question->max_score) ? $question->max_score : 0,
       );
     }
@@ -1075,7 +1075,7 @@ function _quiz_add_questions_to_form(&$form, &$questions, &$quiz, &$question_typ
         '#value' => isset($question->max_score) ? $question->max_score : 0,
       );
     }
-    
+
     $form[$fieldset]['auto_update_max_scores'][$id] = array(
       '#type' => 'checkbox',
       '#default_value' => isset($question->auto_update_max_score) ? $question->auto_update_max_score : 0,
@@ -1533,7 +1533,7 @@ function _quiz_update_items($quiz, $weight_map, $max_scores, $auto_update_max_sc
     }
     else {
       $question->state = QUESTION_ALWAYS;
-    }     
+    }
     $question->weight = $weight;
     $question->max_score = $max_scores[$id];
     $question->auto_update_max_score = $auto_update_max_scores[$id];

